synapse/lib/cmdr.py

```diff
@@ -9,36 +20,595 @@
+        '''
+        self.prox.off('syn:cliserver:outp:%s' % self.session_iden, self._onPrintf)
+        self._connectToServer()
+        self.prox.on('syn:cliserver:outp:%s' % self.session_iden, self._onPrintf)
+        self.printf('Reconnected to server.')
+        self.fire('syn:cliproxy:reconnect', session_iden=self.session_iden)
+
+    def _connectToServer(self):
+        '''
+        Perform the connection, setting the syn:cliserver:init:print:<iden> and
+        syn:cliserver:init:prompt:<iden> event handlers for the duration of the
+        client startup.
+        '''
+        # The CliServer may print events to us during session connection
+        # that we want to print to users.
+        with self.prox.onWith('syn:cliserver:init:print:%s' % self.iden, self._onPrintf):
+            # The syn:cliserver:init:prompt:<iden> events may cause a call/response
+            # cycle with the CliProxy
+            with self.prox.onWith('syn:cliserver:init:prompt:%s' % self.iden,
+                                 self._onInitPrompt):
+                # Do our sessions setup on the server
+                self.session_iden = self.prox.getCmdrSession(self.iden)
+
+    def _cliProxFini(self):
+        self.prox.fini()
+
+    def _onPrintf(self, mesg):
+        msg = mesg[1].get('mesg')
+        self.printf(msg, False)
+
+    def printf(self, mesg, addnl=True):
+        self.outp.printf(mesg, addnl=addnl)
+
+    def _onInitPrompt(self, mesg):
+        '''
+        Print a message and get a user response
+        '''
+        prompt = mesg[1].get('prompt', self.cmdprompt)
+        pmesg = mesg[1].get('mesg')
+        self.printf(pmesg)
+        line = self.get_input(prompt)
+        self.prox.fire('syn:cliserver:init:response:%s' % self.iden, line=line)
+
+    def get_input(self, prompt=None):
+        '''
+        Get the input string to parse.
+
+        Args:
+            prompt (str): Optional string to use as the prompt. Otherwise self.cmdprompt is used.
+
+        Notes:
+            This fires a 'cli:getinput' event prior to getting the
+            input string from the user.
+
+        Returns:
+            str: A string to process.
+        '''
+        return s_cli.Cli.get_input(self, prompt)
+
+    def runCmdLine(self, line):
+        '''
+        Run a single command line on the remote CLI.
+
+        Args:
+            line (str): Line to execute.
+
+        Examples:
+            Execute the 'woot' command with the 'help' switch:
+
+                cli.runCmdLine('woot --help')
+
+        Notes:
+            This generates a random guid and sets the proxy event
+            'syn:cliserver:outp:<guid>' to call self._onPrintf() in order
+            to capture the output sent from the remote server.
+
+        Returns:
+            object: Arbitrary data from the cmd class.
+        '''
+
+        self.fire('cli:cmd:run', line=line)
+
+        iden = s_common.guid()
+        evtname = 'syn:cliserver:outp:%s' % iden
```

Comment:
I wonder if we want to make the request iden from ``s_common.guid((self.session_iden, s_common.guid()))`` to reduce the chance of a collision occuring

Reply:
I think it's fine